testsuite/cluster-test/src/cluster_swarm/cluster_swarm_kube.rs

```diff
@@ -0,0 +1,297 @@
+            let pod_api = Api::v1Pod(self.client.clone()).within(DEFAULT_NAMESPACE);
+            let pod_name = pod_name.to_string();
+            Box::pin(async move {
+                match pod_api.get(&pod_name).await {
+                    Ok(o) => {
+                        let node_name = o.spec.node_name.ok_or_else(|| {
+                            format_err!("node_name not found for pod {}", pod_name)
+                        })?;
+                        let pod_ip = o
+                            .status
+                            .ok_or_else(|| format_err!("status not found for pod {}", pod_name))?
+                            .pod_ip
+                            .ok_or_else(|| format_err!("pod_ip not found for pod {}", pod_name))?;
+                        if node_name.is_empty() || pod_ip.is_empty() {
+                            bail!("node_name not found for pod {}", pod_name)
+                        } else {
+                            Ok((node_name, pod_ip))
+                        }
+                    }
+                    Err(e) => bail!("pod_api.get failed for pod {} : {:?}", pod_name, e),
+                }
+            })
+        })
+        .await
+    }
+
+    async fn delete_pod(&self, name: &str) -> Result<()> {
+        debug!("Deleting Pod {}", name);
+        let pod_api = Api::v1Pod(self.client.clone()).within(DEFAULT_NAMESPACE);
+        pod_api.delete(name, &Default::default()).await?;
+        retry::retry_async(retry::fixed_retry_strategy(2000, 30), || {
+            let pod_api = pod_api.clone();
+            let name = name.to_string();
+            Box::pin(async move {
+                match pod_api.get(&name).await {
+                    Ok(_) => {
+                        bail!("Waiting for pod {} to be deleted..", name);
+                    }
+                    Err(kube::Error::Api(ae)) => {
+                        if ae.code == ERROR_NOT_FOUND {
+                            Ok(())
+                        } else {
+                            bail!("Waiting for pod to be deleted..")
+                        }
+                    }
+                    Err(_) => bail!("Waiting for pod {} to be deleted..", name),
+                }
+            })
+        })
+        .await
+        .map_err(|e| format_err!("Failed to delete pod {}: {:?}", name, e))
+    }
+}
+
+#[async_trait]
+impl ClusterSwarm for ClusterSwarmKube {
+    async fn upsert_validator(
+        &self,
+        index: u32,
+        num_validators: u32,
+        num_fullnodes: u32,
+        image_tag: &str,
+        delete_data: bool,
+    ) -> Result<()> {
+        let pod_name = format!("validator-{}", index);
+        let pod_api = Api::v1Pod(self.client.clone()).within(DEFAULT_NAMESPACE);
+        if pod_api.get(&pod_name).await.is_ok() {
+            self.delete_pod(&pod_name).await?;
+        }
+        let node_name = if let Some(instance) = self.validator_to_node.lock().await.get(&index) {
```

Comment:
node_name is better to be Option<String> instead of String with distinctive values of `""` and `not ""`

Reply:
For the usage, here I would prefer `node_name` to be a String because based on how it gets used in a couple of cases below, the `None` state is the same as `""`. Either we map them to a single state here, or we would have to do it multiple times.

`if node_name.is_empty() {` would become `if node_name.is_none() or node_name.unwrap().is_empty() {` and `validator_spec()` would also need to map `None` -> `""`